### DISCUSSION TITLE
How to set node visibility to hidden with script?

### DISCUSSION THREAD

**Discussion by amariusz** [freeplane/freeplane#1274] [Help] [answered]
I'm wondering if there's a way to programmatically toggle node visibility on and off, the same way as `View->Hidden node` works.
It results in adding `NodeVisibility="HIDDEN"` attribute in the mm file.

I've tried searching in `Help->Freeplane API` but without luck. There's https://docs.freeplane.org/api/org/freeplane/api/NodeRO.html#isVisible()

Is it not available in the API?

**Comment by macmarrum** [comment, marked as answer]
It's missing in the Scripting API, but it can be invoked from a script on selected nodes:
```groovy
menuUtils.executeMenuItems(['NodeEnumerationAction.NodeVisibility.HIDDEN'])
```

**Reply by amariusz** [reply]
Thanks!

**Comment by p-daalder** [comment]
hi, 

Is there a full list of options available in NodeEnumerationAction?
I'm searching for the menu action "UnfoldAll"

Thanks for your reply

Kind regards

Peter

**Reply by HaaiHenkie** [reply]
These are strings for menu items. You can find their names by using the devtools addon: choose from menu: Tools > Developer tools > Menu item info

**Reply by p-daalder** [reply]
> These are strings for menu items. You can find their names by using the devtools addon: choose from menu: Tools > Developer tools > Menu item info

hi 

Thank you for your reply. I didn't have this option ("Developer Tools") initially. After installing the addon ([Development Tools Addon](https://docs.freeplane.org/getting-started/Add-ons_(install).html#for-developers)) the menu item became available.

Unfortunately for me it didn't work. I'm on 1.12.6. I'm getting a null exception.
![image](https://github.com/user-attachments/assets/d24eb377-6897-4ae0-a1fd-96fb926ca3da)

**Reply by p-daalder** [reply]
In the log file I see that "Freeplane\1.12.x\addons\devtools\scripts\menuItemInfo.groovy" is failing

```
Freeplane\1.12.x\addons\devtools\scripts\menuItemInfo.groovy: 15: unable to resolve class org.freeplane.core.ui.MenuBuilder
 @ line 15, column 1.
   import org.freeplane.core.ui.MenuBuilder
   ^
```

**Reply by HaaiHenkie** [reply]
Did you restart Freeplane after installing the addon devtools? For me it is working fine on Freeplane 1.12.6

The menu key for UnfoldAll is: `UnfoldAllAction`

So to execute this menu item from script:

~~~groovy
menuUtils.executeMenuItems(['UnfoldAllAction'])
~~~

**Reply by HaaiHenkie** [reply]
This script will give you an overview of all menu item keys (more than 3500):

~~~groovy
def menuTree = menuUtils.createMenuEntryTree("main_menu")

collectMenuItemKeys(menuTree)

void collectMenuItemKeys(def menuTree) {
    menuTree.children().each {
        def indent = "  " * (it.level - 1)
        if (it.isLeaf()) {
            def menuEntry = it.userObject
            println indent + "'" + it.toString() + "' has menu item key: " + menuEntry.key
        } else {
            println indent + "Menu item keys under: '" + it.toString() +  "'"
            collectMenuItemKeys(it)
        }
    }
}
~~~

**Reply by p-daalder** [reply]
Wow, big thanks @HaaiHenkie. I created a custom script "Menu Tree" with your code and it unveiled all settings.
![image](https://github.com/user-attachments/assets/53853895-111f-4978-8f2e-214c09448aff)


```
...
STDOUT: 'Fold one level' has menu item key: FoldOneLevelAction
STDOUT: 'Unfold all' has menu item key: UnfoldAllAction
STDOUT: 'Fold all' has menu item key: FoldAllAction
STDOUT: 'Keep unfolded nodes' has menu item key: SetAlwaysUnfoldedNodeFlagsAction
...
```
`

I found the action I needed and implemented it in my script to achieve the desired result. My objective was to build a small script to hide all nodes (in a branch). I use this to hide those nodes that can be "archived" ;-). I don't want these to be visible as I don't have to think about them, but I don't want to delete as often people raise questions and I like to be able to consult what was agreed.

```
def bFolded = node.isFolded()

menuUtils.executeMenuItems(['UnfoldAllAction'])

c.selectBranch(node)

menuUtils.executeMenuItems(['NodeEnumerationAction.NodeVisibility.HIDDEN'])

node.setFolded(bFolded)

c.select(node)
```
